modules/perception/inference/libtorch/torch_det_net.cc

```diff
@@ -0,0 +1,108 @@
+/******************************************************************************
+ * Copyright 2020 The Apollo Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#include "modules/perception/inference/libtorch/torch_det_net.h"
+
+#include "cyber/common/log.h"
+
+namespace apollo {
+namespace perception {
+namespace inference {
+
+using apollo::perception::base::Blob;
+
+TorchDetNet::TorchDetNet(const std::string &net_file,
+      const std::string &model_file, const std::vector<std::string> &outputs)
+    : net_file_(net_file), model_file_(model_file), output_names_(outputs) {}
+
+bool TorchDetNet::Init(const std::map<std::string, std::vector<int>> &shapes) {
+  if (gpu_id_ >= 0) {
+    device_type_ = torch::kCUDA;
+    device_id_ = gpu_id_;
+  } else {
+    device_type_ = torch::kCPU;
+  }
+
+  // Init net
+  torch::Device device(device_type_, device_id_);
+  net_ = torch::jit::load(model_file_, device);
+
+  for (auto name : output_names_) {
+    auto blob = std::make_shared<Blob<float>>(2, 6, 1, 1);
+    blobs_.emplace(name, blob);
+  }
+
+  for (auto name : input_names_) {
+    auto iter = shapes.find(name);
+    if (iter != shapes.end()) {
+      auto blob = std::make_shared<Blob<float>>(iter->second);
+      blobs_.emplace(name, blob);
+    }
+  }
+  return true;
+}
+
+TorchDetNet::TorchDetNet(const std::string &net_file,
+                         const std::string &model_file,
+                         const std::vector<std::string> &outputs,
+                         const std::vector<std::string> &inputs)
+    : net_file_(net_file),
+      model_file_(model_file),
+      output_names_(outputs),
+      input_names_(inputs) {}
+
+std::shared_ptr<Blob<float>> TorchDetNet::get_blob(
+    const std::string &name) {
+  auto iter = blobs_.find(name);
+  if (iter == blobs_.end()) {
+    return nullptr;
+  }
+  return iter->second;
+}
+
+void TorchDetNet::Infer() {
+  torch::Device device(device_type_, device_id_);
+  auto blob = blobs_[input_names_[0]];
+  auto input_param = blobs_[input_names_[1]];
+
+  torch::Tensor tensor_image = torch::from_blob(
+                              blob->data()->mutable_cpu_data(),
```

Comment:
If you just need to read cpu data, you don't need `mutable_cpu_data` which is usually used to write data. There should be a function just named `cpu_data()`.

Reply:
I tried cpu_data( ) function but torch::from_blob need void* as initializing argument.

Reply:
Got it.